int32[] ids
int32[] xs
int32[] ys
float32[] thetas